### ISSUE TITLE
`oneOf` with coercion produces the wrong coerced value

### ISSUE THREAD

**Issue by timvandam** [ajv-validator/ajv#2370]
TL;DR:

AJV coercion causes an input that matches a schema to be mutated to an input that does not match.

When using AJV with coercion, it will try to coerce to every type in a `oneOf` to ensure only one type matches. But after doing this for all types in the `oneOf` it does not restore the value to the (coerced) value of whatever the matched type was. My example below shows how a valid input is turned into an invalid input after validation with coercion.

**What version of Ajv are you using? Does the issue happen if you use the latest version?**

v8.12.0

**Ajv options object**

```javascript
{
    allErrors: true,
    coerceTypes: 'array',
    strictSchema: false,
    formats: fullFormats,
}
```

**JSON Schema**

```json
{
        "$schema": "http://json-schema.org/draft-07/schema#",
        "type": "object",
        "properties": {
                "test": {
                        "oneOf": [
                                { "const": null },
                                { "type": "string", "minLength": 1 }
                        ]
                }
        }
}
```

**Sample data**

```json
{ "test": null }
```

**Your code**

```javascript
const Ajv = require('ajv');

const ajv = new Ajv({allErrors:true,coerceTypes:'array',strictSchema:false})

const schema = {
	"$schema": "http://json-schema.org/draft-07/schema#",
	"type": "object",
	"properties": {
		"test": {
			"oneOf": [
				{ "const": null },
				{ "type": "string", minLength: 1 }
			]
		}
	}
};

const validator = ajv.compile(schema);

const input1 = { "test": null };
const input2 = { "test": "test" };

const result1 = validator(input1);
const result2 = validator(input2);

console.log(result1, input1);
console.log(result2, input2);

```

**Validation result, data AFTER validation, error messages**

```javascript
// $ node test.js
true { test: '' } // this no longer matches the schema!
true { test: 'test' }
```

**What results did you expect?**

``javascript
// $ node test.js
true { test: null }
true { test: 'test' }
``

**Are you going to resolve the issue?**
I can try

**Comment by timvandam** [comment]
I see that this is intentional

> Even if it succeeds, Ajv will not backtrack, so you'll get the type of the final coercion even if that's not the one that allowed the data to pass validation. If possible, structure your schema with anyOf, which won't validate subsequent subschemas as soon as it encounters one subschema that matches.

https://ajv.js.org/coercion.html

Is this something that could be added to ajv? Or no chance? Willing to give it a try if so

**Comment by jasoniangreen** [comment]
The advice I have always had around JSON Schema's `oneOf` is that you probably want to use `anyOf`, and that probably applies doubly so when using `coerceTypes`. I doubt there is any will to change this or add another option. As you know, every option compounds the combinatorial complex and footprint of the code base, so there would have to be a very compelling reason.
